File: scripts/lint_tasks.py
import re
import sys
from pathlib import Path

REQUIRED_SECTIONS = ["Description", "Goals", "Requirements", "Subtasks"]
STATUS_TAGS = {"#IceBox", "#Accepted", "#Ready", "#Todo", "#InProgress", "#Done"}


def check_file(path: Path):
    text = path.read_text()
    errors = []
    for section in REQUIRED_SECTIONS:
        pattern = re.compile(
            rf"^#+\s*{re.escape(section)}", re.IGNORECASE | re.MULTILINE
        )
        if not pattern.search(text):
            errors.append(f"Missing section: {section}")
    if not any(tag in text for tag in STATUS_TAGS):
        errors.append("Missing status hashtag")
    if errors:
        return path, errors
    return None


def main() -> None:
    directory = Path(sys.argv[1]) if len(sys.argv) > 1 else Path("docs/agile/tasks")
    failures = []
    for md_file in sorted(directory.glob("*.md")):
        result = check_file(md_file)
        if result:
            failures.append(result)
    if failures:
        for path, errs in failures:
            for err in errs:
                print(f"{path}: {err}")
        sys.exit(1)
    print("All task files have required sections and status hashtags.")


if __name__ == "__main__":
    main()

File: scripts/lint_tasks.ts
import { readFileSync, readdirSync } from 'node:fs';
import path from 'node:path';

const REQUIRED_SECTIONS = ["Description", "Goals", "Requirements", "Subtasks"];
const STATUS_TAGS = new Set(["#IceBox", "#Accepted", "#Ready", "#Todo", "#InProgress", "#Done"]);

function escapeRegExp(str: string): string {
  return str.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function checkFile(filePath: string): string[] {
  const text = readFileSync(filePath, "utf8");
  const errors: string[] = [];
  for (const section of REQUIRED_SECTIONS) {
    const pattern = new RegExp(`^#+\\s*${escapeRegExp(section)}`, "im");
    if (!pattern.test(text)) {
      errors.push(`Missing section: ${section}`);
    }
  }
  if (![...STATUS_TAGS].some((tag) => text.includes(tag))) {
    errors.push("Missing status hashtag");
  }
  return errors;
}

function main(): void {
  const dirArg = process.argv[2];
  const directory = dirArg ? path.resolve(dirArg) : path.resolve("docs/agile/tasks");

  const files = readdirSync(directory).filter((f) => f.endsWith(".md")).sort();
  const failures: [string, string[]][] = [];
  for (const file of files) {
    const filePath = path.join(directory, file);
    const errors = checkFile(filePath);
    if (errors.length > 0) {
      failures.push([filePath, errors]);
    }
  }

  if (failures.length > 0) {
    for (const [file, errs] of failures) {
      for (const err of errs) {
        console.log(`${file}: ${err}`);
      }
    }
    process.exit(1);
  }

  console.log("All task files have required sections and status hashtags.");
}

main();
